client/src/components/form.js

```diff
@@ -0,0 +1,173 @@
+import {
+  Typography,
+  Container,
+  CssBaseline,
+  makeStyles,
+  Button,
+  Divider,
+} from "@material-ui/core";
+
+import { Link } from "react-router-dom";
+import TextInput from "./TextInput";
+
+const useStyles = makeStyles((theme) => ({
+  paper: {
+    marginTop: theme.spacing(8),
+    display: "flex",
+    flexDirection: "column",
+    alignItems: "center",
+  },
+  form: {
+    marginTop: theme.spacing(1),
+    width: "100%",
+    display: "flex",
+    flexDirection: "column",
+    alignItems: "center",
+    justifyContent: "center",
+  },
+  submit: {
+    marginTop: theme.spacing(4),
+    backgroundColor: "#69E781",
+    color: "white",
+    height: "4em",
+    width: "50%",
+  },
+  description: {
+    margin: theme.spacing(0.5, 3, 4),
+  },
+  divider: {
+    backgroundColor: "#69E781",
+    height: "0.25em",
+    width: "8vh",
+    margin: "3em",
+  },
+}));
+
+const Form = ({
+  formTitle,
+  formDesc,
+  redirectText,
+  formLink,
+  buttonText,
+  formName,
+}) => {
+  const [name, setName] = useState({ value: "", error: "" });
+  const [email, setEmail] = useState({ value: "", error: "" });
+  const [password, setPassword] = useState({ value: "", error: "" });
+  const [confirmPassword, setConfirmPassword] = useState({
+    value: "",
+    error: "",
+  });
+
+  const classes = useStyles();
+
+  // Input field validation
+  const validateFields = () => {
+    let isValid = true;
+
+    if (formName === "signup" && name.value === "") {
+      setName({ ...name, error: "Name is required" });
+      isValid = false;
+    }
+
+    if (
+      !/^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$/.test(
+        email.value
+      )
+    ) {
+      setEmail({ ...email, error: "Invalid Email" });
+      isValid = false;
+    }
+    if (password.value.length < 6) {
+      setPassword({ ...password, error: "Invalid Password" });
+      isValid = false;
+    }
+    if (formName === "signup" && password.value !== confirmPassword.value) {
+      setConfirmPassword({
+        ...confirmPassword,
+        error: "Passwords don't match",
+      });
+      isValid = false;
+    }
+    return isValid;
+  };
+
+  // on Button Submit
+  const handleSubmit = (e) => {
+    e.preventDefault();
+    const isValid = validateFields();
+
+    if (isValid) {
+      console.log("Submitted"); // Will replace with API call
+    }
+  };
+
+  return (
+    <Container component="main" maxWidth="xs">
+      <CssBaseline />
+      <div className={classes.paper}>
+        <Typography component="h1" variant="h3">
+          {formTitle}
+        </Typography>
+        <Divider className={classes.divider} />
+        <Typography component="h6" className={classes.description}>
+          {formDesc}{" "}
+          <Link style={{ color: "black" }} to={`/${formLink}`}>
+            {redirectText}
+          </Link>
+        </Typography>
+        <form className={classes.form} noValidate>
+          {formName === "signup" && (
```

Comment:
I like how you wanted to avoid repeated code by combining both the signup and login forms into one component. However, this may be a bit cumbersome as we add more detail to the Login/Signup pages. I don't expect you to change this as it'll be timely, but I would have recommended a few other approaches:

1. Create a component that just wraps the content around the Signup and Login forms (using the children props to have a variable data inside), and then write custom code for the Signup and Login components (all the textfields, etc.)
2. Generalize this component a bit more by allowing users to pass in the text field types for the component. Validation and state management though should happen still in the Signup and Login pages. With reusable components, you want to actual avoid that type of logic, as you will notice, it is hard to generalize and you find yourself writing a lot of if statements.